net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+			rc = libfab_fab_ep_find(tm, &en, cd->fcd_straddr, &ep);
+			if (rc == FI_SUCCESS) {
+				rc = libfab_conn_accept(ep, tm, entry.info);
+				if (rc != FI_SUCCESS)
+					M0_LOG(M0_ERROR, 
+					       "Conn accept failed %d", rc);
+			}
+			fi_freeinfo(entry.info);
+		}
+	} else if (rc == -FI_EAVAIL) {
+		memset(&eq_err, 0, sizeof(eq_err));
+		rc = fi_eq_readerr(eq, &eq_err, 0);
+		if (rc != sizeof(eq_err)) {
+			M0_LOG(M0_ERROR, "fi_eq_readerr returns error =%s",
+			       fi_strerror((int) -rc));
+		} else {
+			M0_LOG(M0_ERROR, "fi_eq_readerr provider err no %d:%s",
+				eq_err.prov_errno,
+				fi_eq_strerror(eq, eq_err.prov_errno,
+					       eq_err.err_data, NULL, 0));
+		}
+	}
+	/* For all other events, there is no error info available.
+	   Hence, all such events can be ignored.
+	*/
+	return 0;
+}
+
+/**
+ * Check connection established and shutdown events for transmit endpoint.
+ * All other types of events are ignored.
+ */
+static void libfab_txep_event_check(struct m0_fab__ep *txep,
+				    struct m0_fab__active_ep *aep,
+				    struct m0_fab__tm *tm)
+{
+	uint32_t rc;
+
+	rc = libfab_check_for_event(aep->aep_tx_res.ftr_eq);
+	if (rc == FI_CONNECTED) {
+		libfab_pending_bufs_send(txep);
+		aep->aep_tx_state = FAB_CONNECTED;
+	} else if (rc == FI_SHUTDOWN) {
+		/* Reset and reopen endpoint */
+		libfab_txep_init(aep, tm, txep);
+	}
+	/* All other types of events can be ignored */
+}
+
+/**
+ * Check for completion events on the completion queue for the receive endpoint
+ */
+static void libfab_rxep_comp_read(struct fid_cq *cq, struct m0_fab__ep *ep)
+{
+	struct m0_fab__buf *buf[FAB_MAX_COMP_READ];
+	m0_bindex_t         len[FAB_MAX_COMP_READ];
+	uint64_t            data[FAB_MAX_COMP_READ];
+	int                 i;
+	int                 cnt;
+
+	memset(&buf, 0, sizeof(buf));
+	memset(&len, 0, sizeof(len));
+	if (cq != NULL) {
+		cnt = libfab_check_for_comp(cq, buf, len, data);
+		for (i = 0; i < cnt; i++) {
+			if (buf[i] != NULL) {
+				if (buf[i]->fb_length == 0)
+					buf[i]->fb_length = len[i];
+				buf[i]->fb_ev_ep = ep;
+				libfab_buf_done(buf[i], 0);
+			}
+			if (data[i] != 0)
+				libfab_buf_done((struct m0_fab__buf*)data[i],
+						0);
+		}
+	}
+}
+
+/**
+ * Check for completion events on the completion queue for the transmit endpoint
+ */
+static void libfab_txep_comp_read(struct fid_cq *cq)
+{
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__buf       *buf[FAB_MAX_COMP_READ];
+	int                       i;
+	int                       cnt;
+
+	memset(&buf, 0, sizeof(buf));
+	cnt = libfab_check_for_comp(cq, buf, NULL, NULL);
+	for (i = 0; i < cnt; i++) {
+		if (buf[i] != NULL) {
+			aep = libfab_aep_get(buf[i]->fb_txctx);
+			if (M0_IN(buf[i]->fb_nb->nb_qtype,
+				  (M0_NET_QT_ACTIVE_BULK_RECV,
+				   M0_NET_QT_ACTIVE_BULK_SEND)))
+				--aep->aep_bulk_cnt;
+			buf[i]->fb_wr_comp_cnt++;
+			if (buf[i]->fb_wr_comp_cnt >= buf[i]->fb_wr_cnt) {
+				libfab_target_notify(buf[i], aep);
```

Comment:
This notify calls fi_send and asserting if send returns error.
_static int libfab_target_notify(struct m0_fab__buf *buf,
				struct m0_fab__active_ep *aep)
{
	uint64_t dummy[2]; /* Dummy data */
	int      ret = 0;
	
	if (buf->fb_nb->nb_qtype == M0_NET_QT_ACTIVE_BULK_RECV) {
		dummy[0] = FAB_DUMMY_DATA;
		dummy[1] = buf->fb_rbd->fbd_bufptr;
		ret = fi_send(aep->aep_txep, &dummy, sizeof(dummy), NULL,
			      0, NULL);
		M0_ASSERT(ret == FI_SUCCESS);
	}

	return M0_RC(ret);
}_

Need to handle this back pressure gracefully e.g. if this function returns error add it to add it to ftm->ftm_done queue and when poll thread handles this it will attempt to do fi_send again, but need to keep state it send is done.

Reply:
Added code for handling backpressure